sdk/lib/drivers/sptilib/sptilib.c

```diff
@@ -0,0 +1,1040 @@
+/*
+ * PROJECT:     ReactOS Storage Stack
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     ATA and SCSI Pass Through Interface for storage drivers
+ * COPYRIGHT:   Copyright 2025 Dmitry Borisov <di.sean@protonmail.com>
+ */
+
+/*
+ * This library provides helper code for handling ATA and SCSI Pass Through IOCTLs.
+ * Typically, these IRPs come from user-mode applications.
+ * The handler will translate the IOCTL into an SRB and send it to the PDO.
+ */
+
+/* INCLUDES *******************************************************************/
+
+#include <ntddk.h>
+#include <ntintsafe.h>
+#include <pseh/pseh2.h>
+#include <scsi.h>
+#include <ntddscsi.h>
+
+#define NDEBUG
+#include <debug.h>
+
+#include "sptilib.h"
+#include "sptilibp.h"
+
+/* PRIVATE FUNCTIONS **********************************************************/
+
+_At_(IrpContext->Srb.SenseInfoBuffer, __drv_freesMem(Mem))
+static
+CODE_SEG("PAGE")
+VOID
+SptiFreeIrpContext(
+    _In_opt_ __drv_freesMem(Mem) PPASSTHROUGH_IRP_CONTEXT IrpContext)
+{
+    PIRP Irp;
+
+    PAGED_CODE();
+
+    if (!IrpContext)
+        return;
+
+    Irp = IrpContext->Irp;
+    if (Irp)
+    {
+        if (Irp->MdlAddress)
+        {
+            MmUnlockPages(Irp->MdlAddress);
+            IoFreeMdl(Irp->MdlAddress);
+        }
+        Irp->MdlAddress = NULL;
+        IoFreeIrp(Irp);
+    }
+
+    if (IrpContext->Srb.SenseInfoBuffer)
+        ExFreePoolWithTag(IrpContext->Srb.SenseInfoBuffer, TAG_SPTI);
+
+    ExFreePoolWithTag(IrpContext, TAG_SPTI);
+}
+
+static IO_COMPLETION_ROUTINE SptiCompletion;
+static
+NTSTATUS
+NTAPI
+SptiCompletion(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _In_ PIRP Irp,
+    _In_reads_opt_(_Inexpressible_("varies")) PVOID Context)
+{
+    UNREFERENCED_PARAMETER(DeviceObject);
+
+    if (Irp->PendingReturned)
+        KeSetEvent(Context, IO_NO_INCREMENT, FALSE);
+
+    return STATUS_MORE_PROCESSING_REQUIRED;
+}
+
+static
+CODE_SEG("PAGE")
+NTSTATUS
```

Comment:
What about making the function return here the `PPASSTHROUGH_IRP_CONTEXT`, instead of an NTSTATUS value?
(like e.g. ExAllocatePool, etc.)